File: static/js/location_service.js
// GPS Location Service for Smart Attendance System
// Ensures faculty can only mark attendance from within college premises

class LocationService {
    constructor() {
        // Vemu Institute of Technology GPS coordinates
        // Located at P. Kothakota, Tirupati-Chittoor Highway, Chittoor District
        this.collegeCoordinates = {
            latitude: 13.4182,   // Center of VEMU boundary coordinates
            longitude: 79.1158,  // Center of VEMU boundary coordinates  
            name: "Vemu Institute of Technology, Chittoor"
        };
        
        // Allowed radius in meters (500m = 0.5km coverage area)
        this.allowedRadius = 500;
        
        this.userLocation = null;
        this.locationAllowed = false;
        this.watchId = null;
    }

    // Calculate distance between two GPS coordinates using Haversine formula
    calculateDistance(lat1, lon1, lat2, lon2) {
        const R = 6371e3; // Earth's radius in meters
        const φ1 = lat1 * Math.PI/180;
        const φ2 = lat2 * Math.PI/180;
        const Δφ = (lat2-lat1) * Math.PI/180;
        const Δλ = (lon2-lon1) * Math.PI/180;

        const a = Math.sin(Δφ/2) * Math.sin(Δφ/2) +
                Math.cos(φ1) * Math.cos(φ2) *
                Math.sin(Δλ/2) * Math.sin(Δλ/2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));

        return R * c; // Distance in meters
    }

    // Show loading screen with GPS tracking message
    showLocationLoading() {
        const loadingHTML = `
            <div id="gps-loading" class="position-fixed top-0 start-0 w-100 h-100 d-flex align-items-center justify-content-center" 
                 style="background: rgba(0,0,0,0.9); z-index: 9999;">
                <div class="text-center text-white">
                    <div class="spinner-border text-primary mb-3" role="status" style="width: 3rem; height: 3rem;">
                        <span class="visually-hidden">Loading...</span>
                    </div>
                    <h4><i class="fas fa-map-marker-alt me-2"></i>GPS Tracking</h4>
                    <p class="mb-2">Verifying your location...</p>
                    <p class="text-muted small">Please ensure location services are enabled</p>
                    <div class="mt-3">
                        <div id="location-status" class="badge bg-warning">
                            <i class="fas fa-satellite-dish me-1"></i>Acquiring GPS signal...
                        </div>
                    </div>
                </div>
            </div>
        `;
        document.body.insertAdjacentHTML('beforeend', loadingHTML);
    }

    // Update location status
    updateLocationStatus(message, type = 'warning') {
        const statusElement = document.getElementById('location-status');
        if (statusElement) {
            const iconMap = {
                'warning': 'fa-satellite-dish',
                'success': 'fa-check-circle',
                'danger': 'fa-exclamation-triangle'
            };
            statusElement.className = `badge bg-${type}`;
            statusElement.innerHTML = `<i class="fas ${iconMap[type]} me-1"></i>${message}`;
        }
    }

    // Check if user is within college premises
    async getCurrentLocation() {
        return new Promise((resolve, reject) => {
            if (!navigator.geolocation) {
                reject(new Error('Geolocation is not supported by this browser'));
                return;
            }

            const options = {
                enableHighAccuracy: true,
                timeout: 15000,
                maximumAge: 60000
            };

            this.updateLocationStatus('Getting precise location...', 'warning');

            navigator.geolocation.getCurrentPosition(
                (position) => {
                    this.userLocation = {
                        latitude: position.coords.latitude,
                        longitude: position.coords.longitude,
                        accuracy: position.coords.accuracy
                    };
                    
                    const distance = this.calculateDistance(
                        this.userLocation.latitude,
                        this.userLocation.longitude,
                        this.collegeCoordinates.latitude,
                        this.collegeCoordinates.longitude
                    );

                    this.locationAllowed = distance <= this.allowedRadius;
                    
                    if (this.locationAllowed) {
                        this.updateLocationStatus(`Location verified - ${Math.round(distance)}m from college`, 'success');
                        setTimeout(() => resolve(true), 1000);
                    } else {
                        this.updateLocationStatus(`Outside college premises - ${Math.round(distance)}m away`, 'danger');
                        setTimeout(() => resolve(false), 2000);
                    }
                },
                (error) => {
                    let errorMessage = 'Location access denied';
                    switch(error.code) {
                        case error.PERMISSION_DENIED:
                            errorMessage = 'Location access denied by user';
                            break;
                        case error.POSITION_UNAVAILABLE:
                            errorMessage = 'Location information unavailable';
                            break;
                        case error.TIMEOUT:
                            errorMessage = 'Location request timeout';
                            break;
                    }
                    this.updateLocationStatus(errorMessage, 'danger');
                    setTimeout(() => reject(new Error(errorMessage)), 2000);
                },
                options
            );
        });
    }

    // Remove loading screen
    hideLocationLoading() {
        const loadingElement = document.getElementById('gps-loading');
        if (loadingElement) {
            loadingElement.style.transition = 'opacity 0.5s';
            loadingElement.style.opacity = '0';
            setTimeout(() => loadingElement.remove(), 500);
        }
    }

    // Show location error screen
    showLocationError(message) {
        this.hideLocationLoading();
        
        const errorHTML = `
            <div id="location-error" class="position-fixed top-0 start-0 w-100 h-100 d-flex align-items-center justify-content-center" 
                 style="background: rgba(220, 53, 69, 0.95); z-index: 9999;">
                <div class="text-center text-white p-4">
                    <div class="mb-4">
                        <i class="fas fa-map-marker-alt-slash" style="font-size: 4rem; opacity: 0.8;"></i>
                    </div>
                    <h3>Access Restricted</h3>
                    <p class="lead">${message}</p>
                    <div class="alert alert-light text-dark mt-4">
                        <h6><i class="fas fa-info-circle me-2"></i>Attendance Policy</h6>
                        <p class="mb-2">• Attendance can only be marked from within college premises</p>
                        <p class="mb-2">• Please ensure you are inside the college campus</p>
                        <p class="mb-0">• Contact IT support if you need assistance</p>
                    </div>
                    <button class="btn btn-light mt-3" onclick="location.reload()">
                        <i class="fas fa-sync me-2"></i>Try Again
                    </button>
                </div>
            </div>
        `;
        document.body.insertAdjacentHTML('beforeend', errorHTML);
    }

    // Main function to verify location before allowing access
    async verifyLocation() {
        try {
            this.showLocationLoading();
            
            const isAllowed = await this.getCurrentLocation();
            
            if (isAllowed) {
                this.hideLocationLoading();
                return true;
            } else {
                const distance = this.calculateDistance(
                    this.userLocation.latitude,
                    this.userLocation.longitude,
                    this.collegeCoordinates.latitude,
                    this.collegeCoordinates.longitude
                );
                
                this.showLocationError(`You are currently ${Math.round(distance)} meters away from ${this.collegeCoordinates.name}. Please come to the college campus to mark attendance.`);
                return false;
            }
        } catch (error) {
            this.showLocationError(`Location verification failed: ${error.message}. Please enable location services and try again.`);
            return false;
        }
    }

    // Get location info for debugging
    getLocationInfo() {
        if (!this.userLocation) return null;
        
        const distance = this.calculateDistance(
            this.userLocation.latitude,
            this.userLocation.longitude,
            this.collegeCoordinates.latitude,
            this.collegeCoordinates.longitude
        );
        
        return {
            userLocation: this.userLocation,
            collegeLocation: this.collegeCoordinates,
            distance: Math.round(distance),
            allowed: this.locationAllowed,
            accuracy: this.userLocation.accuracy
        };
    }
}

// Initialize location service
const locationService = new LocationService();

// Auto-verify location when page loads
document.addEventListener('DOMContentLoaded', async function() {
    // Only verify location on login and attendance pages
    const requiresLocation = window.location.pathname.includes('login') || 
                           window.location.pathname.includes('attendance') ||
                           window.location.pathname.includes('mark_attendance') ||
                           window.location.pathname === '/';
    
    if (requiresLocation) {
        const isAllowed = await locationService.verifyLocation();
        
        if (!isAllowed) {
            // Disable all form submissions if location is not allowed
            const forms = document.querySelectorAll('form');
            forms.forEach(form => {
                form.addEventListener('submit', function(e) {
                    e.preventDefault();
                    alert('Location verification required. Please ensure you are within college premises.');
                });
            });
        }
    }
});

File: location_service.py
import math
import logging

class LocationService:
    def __init__(self):
        # VEMU Institute of Technology boundary coordinates (Updated precise GPS coordinates from map)
        self.college_boundary = [
            (13.417390, 79.116548),    # Point 1 (northeast corner)
            (13.417455, 79.115000),    # Point 2 (northwest corner)
            (13.419012, 79.115183),    # Point 3 (southwest corner)
            (13.418937, 79.116615),    # Point 4 (southeast corner)
        ]
        
        # Calculate circle center and radius from boundary points
        self.college_center = self._calculate_center_point()
        self.college_radius = self._calculate_radius_from_boundary()
    
    def _calculate_center_point(self):
        """
        Calculate the center point of college boundary
        """
        avg_lat = sum(point[0] for point in self.college_boundary) / len(self.college_boundary)
        avg_lon = sum(point[1] for point in self.college_boundary) / len(self.college_boundary)
        return avg_lat, avg_lon
    
    def _calculate_radius_from_boundary(self):
        """
        Calculate the radius needed to encompass all boundary points
        """
        center_lat, center_lon = self.college_center
        max_distance = 0
        
        for point in self.college_boundary:
            distance = self.calculate_distance(center_lat, center_lon, point[0], point[1])
            max_distance = max(max_distance, distance)
        
        # Add 50m buffer to ensure all boundary points are included
        return max_distance + 50
        
    def is_point_in_polygon(self, latitude, longitude):
        """
        Ray-casting algorithm to check if point is inside polygon defined by A,B,C,D boundary points
        """
        x, y = longitude, latitude
        n = len(self.college_boundary)
        inside = False
        
        # Use proper ray-casting algorithm
        j = n - 1
        for i in range(n):
            xi, yi = self.college_boundary[i][1], self.college_boundary[i][0]  # lon, lat
            xj, yj = self.college_boundary[j][1], self.college_boundary[j][0]  # lon, lat
            
            if ((yi > y) != (yj > y)) and (x < (xj - xi) * (y - yi) / (yj - yi) + xi):
                inside = not inside
            j = i
        
        return inside
    
    def calculate_distance(self, lat1, lon1, lat2, lon2):
        """
        Calculate distance between two points in meters using Haversine formula
        """
        R = 6371000  # Earth's radius in meters
        
        lat1_rad = math.radians(lat1)
        lat2_rad = math.radians(lat2)
        delta_lat = math.radians(lat2 - lat1)
        delta_lon = math.radians(lon2 - lon1)
        
        a = (math.sin(delta_lat / 2) * math.sin(delta_lat / 2) +
             math.cos(lat1_rad) * math.cos(lat2_rad) *
             math.sin(delta_lon / 2) * math.sin(delta_lon / 2))
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        
        return R * c
    
    def is_within_college_premises(self, latitude, longitude, buffer_meters=2500):
        """
        Check if user is within college premises using polygon boundary (A,B,C,D points)
        Uses larger buffer for GPS accuracy issues
        """
        try:
            # First check if point is inside the boundary polygon
            if self.is_point_in_polygon(latitude, longitude):
                logging.info(f"Location {latitude}, {longitude} is inside college polygon boundary")
                return True, "Inside college premises"
            
            # If not inside, check if within buffer distance from boundary (increased for GPS accuracy)
            min_distance = float('inf')
            for point in self.college_boundary:
                distance = self.calculate_distance(latitude, longitude, point[0], point[1])
                min_distance = min(min_distance, distance)
            
            # Allow larger buffer zone due to GPS accuracy issues
            if min_distance <= buffer_meters:
                logging.info(f"Location {latitude}, {longitude} is within {buffer_meters}m buffer zone (GPS accuracy adjustment)")
                return True, f"Within college area (GPS accuracy adjusted)"
            
            logging.warning(f"Location {latitude}, {longitude} is {min_distance:.0f}m from college boundary")
            return False, f"You are {min_distance:.0f}m away from college premises"
            
        except Exception as e:
            logging.error(f"Error validating location: {e}")
            return False, "Location validation failed"
    

    
    def is_within_college_circle(self, latitude, longitude):
        """
        Check if user is within the college circular boundary
        """
        try:
            center_lat, center_lon = self.college_center
            distance_from_center = self.calculate_distance(latitude, longitude, center_lat, center_lon)
            
            if distance_from_center <= self.college_radius:
                logging.info(f"Location {latitude}, {longitude} is {distance_from_center:.0f}m from college center (within {self.college_radius:.0f}m radius)")
                return True, f"Inside college premises ({distance_from_center:.0f}m from center)"
            else:
                logging.warning(f"Location {latitude}, {longitude} is {distance_from_center:.0f}m from college center (outside {self.college_radius:.0f}m radius)")
                return False, f"You are {distance_from_center:.0f}m from college center (allowed radius: {self.college_radius:.0f}m)"
                
        except Exception as e:
            logging.error(f"Error checking circle boundary: {e}")
            return False, "Location validation failed"
    
    def get_college_center(self):
        """
        Get the college center coordinates
        """
        return self.college_center
    
    def validate_attendance_location(self, latitude, longitude):
        """
        Main validation function for attendance marking
        """
        if not latitude or not longitude:
            return False, "Location coordinates are required"
        
        try:
            lat = float(latitude)
            lon = float(longitude)
            
            # Validate coordinate ranges
            if not (-90 <= lat <= 90) or not (-180 <= lon <= 180):
                return False, "Invalid GPS coordinates"
            
            return self.is_within_college_premises(lat, lon)
            
        except (ValueError, TypeError):
            return False, "Invalid location data format"

# Initialize location service
location_service = LocationService()